omegaconf/nodes.py

```diff
@@ -163,6 +169,128 @@ def __deepcopy__(self, memo: Dict[int, Any] = {}) -> "AnyNode":
         return res
 
 
+class UnionNode(ValueNode):
+    def __init__(
+        self,
+        value: Any = None,
+        key: Any = None,
+        parent: Optional[Container] = None,
+        is_optional: bool = True,
+        ref_type: Any = Union[Any],
+        *,
+        element_types: List[Any],
+    ):
+        self.element_types = element_types
+        super().__init__(
+            parent=parent,
+            value=value,
+            metadata=Metadata(
+                key=key,
+                optional=is_optional,
+                ref_type=ref_type,
+                object_type=type(value),
+            ),
+        )
+
+    def _wrap_node(
+        self,
+        value: Any,
+        key: Any = None,
+        is_optional: bool = True,
+        parent: Optional[BaseContainer] = None,
+        ref_type: Any = None,
+    ) -> Optional[Node]:
+        from omegaconf.omegaconf import _maybe_wrap
+
+        if value is None:
+            return None
+        else:
+            return _maybe_wrap(
+                ref_type=ref_type,
+                key=key,
+                value=value,
+                is_optional=is_optional,
+                parent=parent,
+            )
+
+    def validate_and_convert(self, value: Any) -> Optional[Any]:
+        from omegaconf._utils import is_container_annotation, is_structured_config
+        from omegaconf.basecontainer import BaseContainer
+
+        if isinstance(value, BaseContainer):
+            ref_type = value._metadata.ref_type
+            value_type = ref_type if is_structured_config(ref_type) else type(value)
+            assert value_type is not None
+            for union_type in self.element_types:
+                if is_container_annotation(value_type) or is_container_annotation(
+                    union_type
+                ):
+                    try:
+                        return self._wrap_node(
+                            value=value._value(), ref_type=union_type
```

Comment:
I'm not sure it's a big deal but I'd rather ask @omry his opinion on this line, because it is changing the original `ref_type` of `value` to `union_type` (not in-place though since we create a new node, this is why it doesn't worry me too much).